// react-pure-carousel/dist/index.es.js
import React from 'react';
import PropTypes from 'prop-types';

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};

  var target = _objectWithoutPropertiesLoose(source, excluded);

  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _toConsumableArray(arr) {
  return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread();
}

function _arrayWithoutHoles(arr) {
  if (Array.isArray(arr)) {
    for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) arr2[i] = arr[i];

    return arr2;
  }
}

function _iterableToArray(iter) {
  if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter);
}

function _nonIterableSpread() {
  throw new TypeError("Invalid attempt to spread non-iterable instance");
}

var addEvent = function addEvent(elem, type, eventHandle) {
  if (elem === null || typeof elem === 'undefined') {
    return;
  }

  if (elem.addEventListener) {
    elem.addEventListener(type, eventHandle, false);
  } else if (elem.attachEvent) {
    elem.attachEvent("on".concat(type), eventHandle);
  } else {
    elem["on".concat(type)] = eventHandle;
  }
};
var removeEvent = function removeEvent(elem, type, eventHandle) {
  if (elem === null || typeof elem === 'undefined') {
    return;
  }

  if (elem.removeEventListener) {
    elem.removeEventListener(type, eventHandle, false);
  } else if (elem.detachEvent) {
    elem.detachEvent("on".concat(type), eventHandle);
  } else {
    elem["on".concat(type)] = null;
  }
};

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = "/* Carousel Component */\n.styles_carousel__1WzHI {\n  position: relative;\n  overflow: hidden;\n  width: 100%;\n  height: 100%;\n  outline: none;\n}\n\n.styles_carousel__1WzHI.styles_light__aWozP {}\n\n.styles_buttonArrow__3mxXi {\n  position: absolute;\n  top: calc((100% - 100px) / 2);\n  width: 100px;\n  height: 100px;\n  background: none !important;\n  outline: none !important;\n  border: 0;\n  box-shadow: none;\n  cursor: pointer;\n  z-index: 10;\n  opacity: 0.7;\n  -webkit-transition: .3s all ease;\n  -o-transition: .3s all ease;\n  transition: .3s all ease;\n}\n\n.styles_buttonArrow__3mxXi:hover {\n  opacity: 1;\n}\n\n.styles_buttonArrow__3mxXi:disabled {\n  opacity: 0.3;\n  cursor: default;\n}\n\n.styles_buttonArrow__3mxXi.styles_buttonArrowLeft__2N5hJ {\n  left: 0;\n}\n\n.styles_buttonArrow__3mxXi.styles_buttonArrowRight__3zBDJ {\n  right: 0;\n}\n\n.styles_buttonArrow__3mxXi.styles_buttonArrowTop__1KNYv {\n  top: 0;\n  left: calc((100% - 100px) / 2);\n}\n\n.styles_buttonArrow__3mxXi.styles_buttonArrowBottom__2QOO5 {\n  top: auto;\n  bottom: 0;\n  left: calc((100% - 100px) / 2);\n}\n\n.styles_arrow__3rhMx {\n  display: inline-block;\n  width: 30px;\n  height: 30px;\n  border-top: 3px solid #FFFFFF;\n  border-right: 3px solid #FFFFFF;\n}\n\n.styles_carousel__1WzHI.styles_dark__Rt22A .styles_arrow__3rhMx {\n  border-color: #333333;\n}\n\n.styles_buttonArrow__3mxXi.styles_buttonArrowLeft__2N5hJ .styles_arrow__3rhMx {\n  transform: rotate(-135deg);\n}\n\n.styles_buttonArrow__3mxXi.styles_buttonArrowRight__3zBDJ .styles_arrow__3rhMx {\n  transform: rotate(45deg);\n}\n\n.styles_buttonArrow__3mxXi.styles_buttonArrowTop__1KNYv .styles_arrow__3rhMx {\n  transform: rotate(-45deg);\n}\n\n.styles_buttonArrow__3mxXi.styles_buttonArrowBottom__2QOO5 .styles_arrow__3rhMx {\n  transform: rotate(135deg);\n}\n\n.styles_pagination__24s15 {\n  position: absolute;\n  bottom: 0;\n  width: 100%;\n  display: block;\n  text-align: center;\n  padding: 20px 15%;\n  box-sizing: border-box;\n  left: 0;\n  z-index: 11;\n}\n\n.styles_pagination__24s15.styles_opposite__3668C {\n  bottom: auto;\n  top: 0;\n}\n\n.styles_carousel__1WzHI.styles_vertical__2hmE6 .styles_pagination__24s15 {\n  width: auto;\n  padding: 15% 20px;\n  height: 100%;\n  display: flex;\n  flex-direction: column;\n  justify-content: center;\n}\n\n.styles_carousel__1WzHI.styles_vertical__2hmE6 .styles_pagination__24s15.styles_opposite__3668C {\n  left: auto;\n  right: 0;\n}\n\n.styles_pagination__24s15 div {\n  height: 4px;\n  max-width: 40px;\n  min-width: 4px;\n  display: inline-block;\n  background-color: #FFFFFF;\n  opacity: 0.5;\n  margin-right: 5px;\n  margin-left: 5px;\n  border-radius: 4px;\n  -webkit-transition: .3s all ease;\n  -o-transition: .3s all ease;\n  transition: .3s all ease;\n}\n\n.styles_carousel__1WzHI.styles_vertical__2hmE6 .styles_pagination__24s15 div {\n  max-height: 40px;\n  min-height: 4px;\n  display: block;\n  margin: 0;\n  margin-top: 5px;\n  margin-bottom: 5px;\n}\n\n.styles_carousel__1WzHI.styles_dark__Rt22A .styles_pagination__24s15 div {\n  background-color: #333333;\n}\n\n.styles_pagination__24s15 div.styles_active__3G4VH {\n  opacity: 1;\n}\n/* END Carousel Component */\n\n/* Item Component */\n.styles_carouselItem__35EUI {\n  width: 100%;\n  text-align: center;\n  height: 100%;\n  position: absolute;\n  display: none;\n  z-index: 10;\n  box-sizing: border-box;\n  background-position: center;\n  background-repeat: no-repeat;\n  background-size: cover;\n  flex-direction: column;\n  justify-content: center;\n}\n\n.styles_carousel__1WzHI.styles_horizontal__2L-77 .styles_carouselItem__35EUI {\n  bottom: 0;\n  padding: 30px 0;\n}\n\n.styles_carousel__1WzHI.styles_vertical__2hmE6 .styles_carouselItem__35EUI {\n  left: 0;\n  padding: 0 30px;\n}\n\n.styles_carouselItem__35EUI.styles_next___68Hz {\n  display: flex;\n  -webkit-transition: 1s all ease;\n  -o-transition: 1s all ease;\n  transition: 1s all ease;\n}\n\n.styles_carousel__1WzHI.styles_horizontal__2L-77 .styles_carouselItem__35EUI.styles_next___68Hz {\n  left: 100%;\n}\n\n.styles_carousel__1WzHI.styles_vertical__2hmE6 .styles_carouselItem__35EUI.styles_next___68Hz {\n  top: 100%;\n}\n\n.styles_carouselItem__35EUI.styles_prev__2OWWh {\n  display: flex;\n  -webkit-transition: 1s all ease;\n  -o-transition: 1s all ease;\n  transition: 1s all ease;\n}\n\n.styles_carousel__1WzHI.styles_horizontal__2L-77 .styles_carouselItem__35EUI.styles_prev__2OWWh {\n  left: -100%;\n}\n\n.styles_carousel__1WzHI.styles_vertical__2hmE6 .styles_carouselItem__35EUI.styles_prev__2OWWh {\n  top: -100%;\n}\n\n.styles_carouselItem__35EUI.styles_active__3G4VH {\n  display: flex;\n  -webkit-transition: 1s all ease;\n  -o-transition: 1s all ease;\n  transition: 1s all ease;\n}\n\n.styles_carousel__1WzHI.styles_horizontal__2L-77 .styles_carouselItem__35EUI.styles_active__3G4VH {\n  left: 0;\n}\n\n.styles_carousel__1WzHI.styles_vertical__2hmE6 .styles_carouselItem__35EUI.styles_active__3G4VH {\n  top: 0;\n}\n\n.styles_carouselItem__35EUI.styles_notAnimated__37LoX {\n  -webkit-transition: none;\n  -o-transition: none;\n  transition: none;\n}\n/* END Item Component */\n\n/* Background Component */\n.styles_carouselBackground__2fa04 {\n  position: absolute;\n  top: 0;\n  left: 0;\n  width: 100%;\n  height: 100%;\n  overflow: hidden;\n  z-index: 9;\n  background-position: center;\n  background-repeat: no-repeat;\n  background-size: cover;\n}\n/* END Background Component */\n";
var styles = {"carousel":"styles_carousel__1WzHI","light":"styles_light__aWozP","buttonArrow":"styles_buttonArrow__3mxXi","buttonArrowLeft":"styles_buttonArrowLeft__2N5hJ","buttonArrowRight":"styles_buttonArrowRight__3zBDJ","buttonArrowTop":"styles_buttonArrowTop__1KNYv","buttonArrowBottom":"styles_buttonArrowBottom__2QOO5","arrow":"styles_arrow__3rhMx","dark":"styles_dark__Rt22A","pagination":"styles_pagination__24s15","opposite":"styles_opposite__3668C","vertical":"styles_vertical__2hmE6","active":"styles_active__3G4VH","carouselItem":"styles_carouselItem__35EUI","horizontal":"styles_horizontal__2L-77","next":"styles_next___68Hz","prev":"styles_prev__2OWWh","notAnimated":"styles_notAnimated__37LoX","carouselBackground":"styles_carouselBackground__2fa04"};
styleInject(css);

var Item = function Item(_ref) {
  var children = _ref.children,
      _ref$className = _ref.className,
      className = _ref$className === void 0 ? null : _ref$className,
      isActive = _ref.isActive,
      isNext = _ref.isNext,
      isPrev = _ref.isPrev,
      isAnimated = _ref.isAnimated,
      _ref$duration = _ref.duration,
      duration = _ref$duration === void 0 ? 1 : _ref$duration,
      backgroundColor = _ref.backgroundColor,
      backgroundImage = _ref.backgroundImage,
      alignItems = _ref.alignItems,
      props = _objectWithoutProperties(_ref, ["children", "className", "isActive", "isNext", "isPrev", "isAnimated", "duration", "backgroundColor", "backgroundImage", "alignItems"]);

  var getStyles = function getStyles() {
    var styles$$1 = {};

    if (backgroundImage) {
      styles$$1 = _objectSpread2({}, styles$$1, {
        backgroundImage: "url(".concat(backgroundImage, ")")
      });
    }

    if (backgroundColor) {
      styles$$1 = _objectSpread2({}, styles$$1, {
        backgroundColor: backgroundColor
      });
    }

    if (duration) {
      styles$$1 = _objectSpread2({}, styles$$1, {
        WebkitTransitionDuration: "".concat(duration, "s"),
        OTransitionDuration: "".concat(duration, "s"),
        transitionDuration: "".concat(duration, "s")
      });
    }

    if (alignItems) {
      var align = {
        'center': alignItems === 'center',
        'flex-start': alignItems === 'top',
        'flex-end': alignItems === 'bottom'
      };
      styles$$1 = _objectSpread2({}, styles$$1, {
        justifyContent: Object.keys(align).find(function (index) {
          return align[index];
        })
      });
    }

    return styles$$1;
  };

  return React.createElement("div", _extends({
    className: "".concat(styles.carouselItem, " ").concat(isActive ? styles.active : '', " ").concat(isNext ? styles.next : '', " ").concat(isPrev ? styles.prev : '', " ").concat(!isAnimated ? styles.notAnimated : '', " ").concat(className || ''),
    style: getStyles()
  }, props), children);
};
Item.propTypes = {
  className: PropTypes.string,
  duration: PropTypes.number,
  isActive: PropTypes.bool.isRequired,
  isNext: PropTypes.bool.isRequired,
  isPrev: PropTypes.bool.isRequired,
  isAnimated: PropTypes.bool.isRequired,
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  backgroundColor: PropTypes.string,
  backgroundImage: PropTypes.string,
  alignItems: PropTypes.string
};
Item.defaultProps = {
  isActive: false,
  isNext: false,
  isPrev: false,
  isAnimated: false,
  duration: 1,
  className: null,
  backgroundColor: null,
  backgroundImage: null,
  alignItems: 'center'
};

var Background = function Background(_ref) {
  var _ref$src = _ref.src,
      src = _ref$src === void 0 ? null : _ref$src,
      _ref$className = _ref.className,
      className = _ref$className === void 0 ? null : _ref$className,
      _ref$color = _ref.color,
      color = _ref$color === void 0 ? null : _ref$color,
      props = _objectWithoutProperties(_ref, ["src", "className", "color"]);

  var getStyles = function getStyles() {
    var styles$$1 = {};

    if (src) {
      styles$$1 = _objectSpread2({}, styles$$1, {
        backgroundImage: "url(".concat(src, ")")
      });
    }

    if (color) {
      styles$$1 = _objectSpread2({}, styles$$1, {
        backgroundColor: color
      });
    }

    return styles$$1;
  };

  return React.createElement("div", _extends({
    className: "".concat(styles.carouselBackground, " ").concat(className || ''),
    style: getStyles()
  }, props));
};
Background.propTypes = {
  className: PropTypes.string,
  src: PropTypes.string,
  color: PropTypes.string
};
Background.defaultProps = {
  className: null,
  src: null,
  color: null
};

var LEFT = 'left';
var RIGHT = 'right';
var TOP = 'top';
var BOTTOM = 'bottom';

var Arrow = function Arrow(_ref) {
  var renderArrow = _ref.renderArrow,
      disabled = _ref.disabled,
      onClick = _ref.onClick,
      _ref$direction = _ref.direction,
      direction = _ref$direction === void 0 ? LEFT : _ref$direction,
      className = _ref.className,
      props = _objectWithoutProperties(_ref, ["renderArrow", "disabled", "onClick", "direction", "className"]);

  var arrowClassName = function arrowClassName() {
    switch (direction) {
      case LEFT:
        return styles.buttonArrowLeft;

      case TOP:
        return styles.buttonArrowTop;

      case BOTTOM:
        return styles.buttonArrowBottom;

      default:
        return styles.buttonArrowRight;
    }
  };

  return React.createElement("button", _extends({
    onClick: onClick,
    type: "button",
    className: "".concat(styles.buttonArrow, " ").concat(arrowClassName(), " ").concat(className || ''),
    disabled: disabled
  }, props), renderArrow, !renderArrow && React.createElement("span", {
    className: styles.arrow
  }));
};
Arrow.propTypes = {
  className: PropTypes.string,
  renderArrow: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  direction: PropTypes.oneOf([LEFT, RIGHT, TOP, BOTTOM])
};
Arrow.defaultProps = {
  className: null,
  renderArrow: null,
  disabled: false,
  onClick: function onClick() {},
  direction: 'left'
};

var Pagination = function Pagination(_ref) {
  var slides = _ref.slides,
      className = _ref.className,
      activeSlide = _ref.activeSlide,
      vertical = _ref.vertical,
      alignOpposite = _ref.alignOpposite,
      props = _objectWithoutProperties(_ref, ["slides", "className", "activeSlide", "vertical", "alignOpposite"]);

  var pageWidth = function pageWidth() {
    var filteredSlides = slides.filter(function (item) {
      return !item.clone;
    });
    return 100 / filteredSlides.length;
  };

  return React.createElement("div", _extends({
    className: "".concat(styles.pagination, " ").concat(alignOpposite ? styles.opposite : '', " ").concat(className || '')
  }, props), slides.map(function (slide, index) {
    return !slide.clone && React.createElement("div", {
      key: index,
      className: index === activeSlide || slides[activeSlide].actualInd === index ? styles.active : '',
      style: vertical ? {
        height: "calc(".concat(pageWidth(), "% - 10px)")
      } : {
        width: "calc(".concat(pageWidth(), "% - 10px)")
      }
    });
  }));
};
Pagination.propTypes = {
  className: PropTypes.string,
  slides: PropTypes.array,
  activeSlide: PropTypes.number,
  vertical: PropTypes.bool,
  alignOpposite: PropTypes.bool
};
Pagination.defaultProps = {
  className: null,
  slides: [],
  activeSlide: 0,
  vertical: false,
  alignOpposite: false
};

var Carousel =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Carousel, _React$Component);

  function Carousel(props) {
    var _this;

    _classCallCheck(this, Carousel);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Carousel).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "timer", null);

    _defineProperty(_assertThisInitialized(_this), "keyCodeMap", null);

    _defineProperty(_assertThisInitialized(_this), "afterSlideCallback", function () {
      var _this$props = _this.props,
          afterSlide = _this$props.afterSlide,
          duration = _this$props.duration;
      var activeSlide = _this.state.activeSlide;
      setTimeout(function () {
        return afterSlide({
          activeSlide: activeSlide
        });
      }, duration * 1000);
    });

    _defineProperty(_assertThisInitialized(_this), "beforeSlideCallback", function (newActiveSlide) {
      var beforeSlide = _this.props.beforeSlide;
      var activeSlide = _this.state.activeSlide;
      beforeSlide({
        activeSlide: activeSlide,
        newActiveSlide: newActiveSlide
      });
    });

    _defineProperty(_assertThisInitialized(_this), "_renderItem", function (ch, index) {
      var duration = _this.props.duration;
      var _this$state = _this.state,
          items = _this$state.items,
          activeSlide = _this$state.activeSlide,
          animatedSlide = _this$state.animatedSlide;
      return React.cloneElement(ch, {
        isPrev: activeSlide === 0 ? index === items.length - 1 : index === activeSlide - 1,
        isNext: activeSlide === items.length - 1 ? index === 0 : index === activeSlide + 1,
        isActive: index === activeSlide,
        isAnimated: index === activeSlide || index === animatedSlide,
        duration: duration
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleFocus", function () {
      var hasFocus = _this.state.hasFocus;

      if (!hasFocus) {
        _this.setState({
          hasFocus: true
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleBlur", function () {
      var hasFocus = _this.state.hasFocus;

      if (hasFocus) {
        _this.setState({
          hasFocus: false
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "bindEvents", function () {
      addEvent(document, 'keydown', _this.handleKeyPress);
    });

    _defineProperty(_assertThisInitialized(_this), "unbindEvents", function () {
      removeEvent(document, 'keydown', _this.handleKeyPress);
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyPress", function (e) {
      var enableKeyboardControls = _this.props.enableKeyboardControls;
      var hasFocus = _this.state.hasFocus;

      if (hasFocus && enableKeyboardControls) {
        var actionName = _this.keyCodeMap[e.keyCode];

        switch (actionName) {
          case 'nextSlide':
            _this.next();

            break;

          case 'previousSlide':
            _this.prev();

            break;

          default:
            break;
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "getKeyCodeMap", function (keyCodeConfig) {
      var keyCodeMap = {};
      Object.keys(keyCodeConfig).forEach(function (actionName) {
        keyCodeConfig[actionName].forEach(function (keyCode) {
          return keyCodeMap[keyCode] = actionName;
        });
      });
      return keyCodeMap;
    });

    _this.state = {
      items: [],
      activeSlide: 0,
      animatedSlide: 0,
      hasFocus: false
    };
    _this.keyCodeConfig = {
      nextSlide: [39, 68, 38, 87],
      previousSlide: [37, 65, 40, 83]
    };
    return _this;
  }

  _createClass(Carousel, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      var _this$props2 = this.props,
          interval = _this$props2.interval,
          duration = _this$props2.duration;
      this.bindEvents();

      if (interval) {
        this.timer = setInterval(function () {
          return _this2.next();
        }, interval + duration * 1000);
      }

      this.keyCodeMap = this.getKeyCodeMap(this.keyCodeConfig);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.unbindEvents();

      if (this.timer) {
        clearInterval(this.timer);
      }
    }
  }, {
    key: "prev",
    value: function prev() {
      var newActiveSlide = this.prevActiveSlide;
      this.beforeSlideCallback(newActiveSlide);
      this.setState(function (prevState) {
        return {
          activeSlide: prevState.activeSlide === 0 ? prevState.items.length - 1 : prevState.activeSlide - 1,
          animatedSlide: prevState.activeSlide
        };
      }, this.afterSlideCallback);
    }
  }, {
    key: "next",
    value: function next() {
      var newActiveSlide = this.nextActiveSlide;
      this.beforeSlideCallback(newActiveSlide);
      this.setState(function (prevState) {
        return {
          activeSlide: prevState.activeSlide === prevState.items.length - 1 ? 0 : prevState.activeSlide + 1,
          animatedSlide: prevState.activeSlide
        };
      }, this.afterSlideCallback);
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props3 = this.props,
          className = _this$props3.className,
          children = _this$props3.children,
          looped = _this$props3.looped,
          prevArrow = _this$props3.prevArrow,
          nextArrow = _this$props3.nextArrow,
          hidePagination = _this$props3.hidePagination,
          afterSlide = _this$props3.afterSlide,
          beforeSlide = _this$props3.beforeSlide,
          vertical = _this$props3.vertical,
          alignPaginationOpposite = _this$props3.alignPaginationOpposite,
          theme = _this$props3.theme,
          duration = _this$props3.duration,
          enableKeyboardControls = _this$props3.enableKeyboardControls,
          props = _objectWithoutProperties(_this$props3, ["className", "children", "looped", "prevArrow", "nextArrow", "hidePagination", "afterSlide", "beforeSlide", "vertical", "alignPaginationOpposite", "theme", "duration", "enableKeyboardControls"]);

      var _this$state2 = this.state,
          items = _this$state2.items,
          activeSlide = _this$state2.activeSlide;

      if (!children) {
        return React.createElement("div", null);
      }

      var itemsWithoutClones = items.filter(function (item) {
        return !item.clone;
      });
      return React.createElement("div", _extends({
        onFocus: this.handleFocus,
        onBlur: this.handleBlur,
        tabIndex: 0,
        className: "".concat(styles.carousel, " ").concat(this.themeClass, " ").concat(vertical ? styles.vertical : styles.horizontal, " ").concat(className || '')
      }, props), Array.isArray(children) && children.filter(function (ch) {
        return ch.type === Background;
      }), React.Children.map(items, function (ch, index) {
        return _this3._renderItem(ch, index);
      }), !hidePagination && React.createElement(Pagination, {
        slides: items,
        activeSlide: activeSlide,
        vertical: vertical,
        alignOpposite: alignPaginationOpposite
      }), items.length > 1 && React.createElement(React.Fragment, null, React.createElement(Arrow, {
        onClick: function onClick() {
          return _this3.prev();
        },
        disabled: !looped && activeSlide === 0,
        renderArrow: prevArrow,
        direction: vertical ? TOP : LEFT
      }), React.createElement(Arrow, {
        onClick: function onClick() {
          return _this3.next();
        },
        disabled: !looped && activeSlide === itemsWithoutClones.length - 1,
        renderArrow: nextArrow,
        direction: vertical ? BOTTOM : RIGHT
      })));
    }
  }, {
    key: "prevActiveSlide",
    get: function get() {
      var _this$state3 = this.state,
          activeSlide = _this$state3.activeSlide,
          items = _this$state3.items;
      return activeSlide === 0 ? items.length - 1 : activeSlide - 1;
    }
  }, {
    key: "nextActiveSlide",
    get: function get() {
      var _this$state4 = this.state,
          activeSlide = _this$state4.activeSlide,
          items = _this$state4.items;
      return activeSlide === items.length - 1 ? 0 : activeSlide + 1;
    }
  }, {
    key: "themeClass",
    get: function get() {
      var theme = this.props.theme;

      if (theme === 'dark') {
        return styles.dark;
      }

      return styles.light;
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      var children = props.children;

      if (children && Array.isArray(children)) {
        var isBackground = children.find(function (ch) {
          return ch.type === Background;
        });
        var itemsArr = children;

        if (isBackground) {
          children.forEach(function (ch) {
            if (Array.isArray(ch)) {
              itemsArr = ch;
            }
          });
        }

        var items = itemsArr.filter(function (ch) {
          return ch.type === Item;
        });

        if (items.length === 2) {
          items = [].concat(_toConsumableArray(items), [_objectSpread2({}, items[0], {
            clone: true,
            actualInd: 0
          }), _objectSpread2({}, items[1], {
            clone: true,
            actualInd: 1
          })]);
        }

        return _objectSpread2({}, state, {
          items: items
        });
      }

      return state;
    }
  }]);

  return Carousel;
}(React.Component);

_defineProperty(Carousel, "Item", Item);

_defineProperty(Carousel, "Background", Background);

Carousel.propTypes = {
  className: PropTypes.string,
  duration: PropTypes.number,
  theme: PropTypes.string,
  interval: PropTypes.number,
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  looped: PropTypes.bool.isRequired,
  prevArrow: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  nextArrow: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  hidePagination: PropTypes.bool,
  afterSlide: PropTypes.func,
  beforeSlide: PropTypes.func,
  vertical: PropTypes.bool,
  alignPaginationOpposite: PropTypes.bool,
  enableKeyboardControls: PropTypes.bool
};
Carousel.defaultProps = {
  className: null,
  duration: 1,
  theme: 'light',
  interval: null,
  children: null,
  looped: true,
  prevArrow: null,
  nextArrow: null,
  hidePagination: false,
  afterSlide: function afterSlide() {},
  beforeSlide: function beforeSlide() {},
  vertical: false,
  alignPaginationOpposite: false,
  enableKeyboardControls: true
};

export { Carousel };
